### ISSUE TITLE
Two finger scrolling does not always work

### ISSUE THREAD

**Issue by martinstark** [mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver#1]
I have both the original Magic Trackpad 2 and the 2024 USB-C version.

If I put two fingers lightly on the 2024 trackpad it does not always result in scrolling and instead moves the cursor around. It will sometimes drop scroll input and start moving the mouse instead while doing two finger scrolling, or trigger a click. I have no issues with the previous generation using the same input settings in Arch/Sway.

```bash
input "76:804:Apple_Inc._Magic_Trackpad_2" {
  tap enabled
  accel_profile adaptive
  pointer_accel 0.25
}
```

**Comment by mr-cal** [comment]
Does increasing pressure increase the likelihood of a two-finger scroll being detected?

I wonder if there are some differences in pressure sensitivity or how much variance there is from device to device.  You may be able to experiment with [ABS_MT_PRESSURE or ABS_PRESSURE](https://www.kernel.org/doc/Documentation/input/multi-touch-protocol.txt) and see if it makes a difference.

https://github.com/mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver/blob/7bf74255ea0dd6a9c0c7790cf4c115b52637e976/linux/drivers/hid/hid-magicmouse.c#L638-L639

**Comment by SeDi343** [comment]
I have the same experience. For me it looks like difference in distance of fingers, instead of pressure. Also 3 fingers multi gestures is rarely working. and 4 fingers is not working at all. This seems like some fine tuning issue.

**Comment by SeDi343** [comment]
I checked this now again, it all works but it seems like as you guessed @mr-cal.

I increased the pressure and it work, seems like there are different `ABS_MT_PRESSURE` and `ABS_PRESSURE` values needed for the USB-C version.

**Comment by mr-cal** [comment]
@SeDi343, thanks for experimenting and testing.  Mind sharing what values worked for you?

**Comment by SeDi343** [comment]
Hey @mr-cal I did not test it yet, since I installed the driver via the deb file, and have secure boot enabled on my companies device. But I suspect the same thing as you, that the pressure sensitivity is lower than on the Lightning Version, since much more force is needed on the USBC Version with this settings. Basically so much force, that you can not smoothly scroll with your fingers and stick on the surface.

**Comment by SeDi343** [comment]
Than lower the number (253) than less pressure is needed for the input? Or how can I read this @mr-cal?

**Comment by mr-cal** [comment]
@SeDi343 ah, I see. As far as I understand, increasing `ABS_MT_PRESSURE` should increase the sensitivity.  It's already pretty high so I'm not sure if it's going to make a noticeable difference.

**Comment by SeDi343** [comment]
What is the max value for `ABS_MT_PRESSURE` and `ABS_PRESSURE`? `255`? Since I found this https://www.linuxquestions.org/questions/linux-kernel-70/what-does-the-fuzz-input_set_abs_params-function-4175577781/ where `MAX_12BIT` is mentioned.

**Comment by SeDi343** [comment]
Also, if it is installed already via dpkg from the downloaded deb file and I now build the deb myself and want to install it via dpkg do i need to uninstall the currently installed one first?

```
$ sudo dpkg -I magicmouse-hid_2.1.0-0.deb 
[sudo] password for admin: 
 new Debian package, version 2.0.
 size 28626 bytes: control archive=903 bytes.
       0 bytes,     0 lines      conffiles            
     461 bytes,    13 lines      control              
    1001 bytes,    12 lines      md5sums              
     188 bytes,    11 lines   *  postinst             #!/bin/bash
     189 bytes,    11 lines   *  prerm                #!/bin/bash
 Package: magicmouse-hid
 Version: 2.1.0-0
 Maintainer: Callahan Kovacs <callahankovacs@gmail.com>
 Description: Magic Mouse 2 driver for Linux
  This driver makes use of DKMS to load the driver in the kernel, while
  making use of udev rules to load the driver when a mouse is connected.
 Homepage: https://github.com/mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver
 Architecture: all
 Depends: dkms, bash
 Section: utils
 Priority: optional
 Essential: no
 Installed-Size: 354
```

**Comment by SeDi343** [comment]
Will experiment a bit in my fork https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver and create PR when I find a proper setting with `ABS_MT_PRESSURE` and `ABS_PRESSURE`

**Comment by SeDi343** [comment]
Guessing, I installed it correctly, since scrolling is more reliable now, do you know, how to make the sensitivity even more sensitive if 255 is used already?

**Comment by SeDi343** [comment]
I think I am doing something wrong, since I see no difference in value `10-255` for `ABS_MT_PRESSURE` and `ABS_PRESSURE`. Whenever I make changes I run `./build-deb.sh` and then `sudo dpkg -I magicmouse-hid_2.1.0-0.deb`. And have Secure Boot active. Do I need to uninstall the driver first before installing it again?

**Comment by SeDi343** [comment]
Any idea what `input_set_events_per_packet(input, 60);` stands for? Does it have something to do with the polling rate?

**Comment by SeDi343** [comment]
Added PR https://github.com/mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver/pull/3 please check, and give feedback. Those changes actually made it more reliable for me.

**Comment by SeDi343** [comment]
I am still confused because `libinput measure touchpad-pressure` prints out 253 as max value

`|  2303  |   x    |  x   |   x    |        |  0   | 253  | 0  | 103  |    82    |`

**Comment by SeDi343** [comment]
A problem I encountered is this using `libinput debug-events`

event258 - Apple Inc. Magic Trackpad 2: kernel bug: Touch jump detected and discarded.
See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details
event258 - Apple Inc. Magic Trackpad 2: kernel bug: Touch jump detected and discarded.
See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details
event258 - Apple Inc. Magic Trackpad 2: kernel bug: Touch jump detected and discarded.
See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details
event258 - Apple Inc. Magic Trackpad 2: WARNING: log rate limit exceeded (5 msgs per 24h). Discarding future messages.

**Comment by SeDi343** [comment]
Would really like to contribute but its not self explaining a lot and documentation is very bad regarding the parameters

**Comment by SeDi343** [comment]
It could be also something orientation of fingers specially with 4 finger gestures i found out its easier to have the hand 90 degrees to the touchpad than 0 degrees

**Comment by SeDi343** [comment]
@martinstark in case u use the touch pad with you left hand using the parameter `left_handed true` might help.

```
input "76:804:Apple_Inc._Magic_Trackpad_USB-C" {
   dwt disabled
   tap disabled
   click_method clickfinger
   natural_scroll enabled
   #accel_profile "flat"
   accel_profile "adaptive"
   pointer_accel 0.5
   left_handed true
   middle_emulation disabled
}
```

**Comment by mr-cal** [comment]
> Would really like to contribute but its not self explaining a lot and documentation is very bad regarding the parameters

Agreed, I find the documentation opaque and it doesn't help that the magic trackpad is a blackbox.  It would be invaluable to see the process the previous devs used to build the original driver.  

> A problem I encountered is this using libinput debug-events
>
> event258 - Apple Inc. Magic Trackpad 2: kernel bug: Touch jump detected and discarded.
> See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details

Interesting, I get this error too.  I wonder if this occurs on older trackpads too.  I went down a rabbit hole on a dmesg error  `unknown main item tag 0x0` only to find logs online where this occurs with previous models.

> I am still confused because libinput measure touchpad-pressure prints out 253 as max value

I see the same thing :upside_down_face:.  Even lowering it doesn't change what is reported by libinput, so I'm not sure whats going on there.  Using evtest, libinput, etc, the trackpad appears extremely sensitive at detecting the lightest pressures of multiple fingers. If anything, I wonder if it could be when libinput decides to go from `POINTER_MOTION` to `POINTER_SCROLL_FINGER` and could be improved with libinput tuning and not at the driver level.

> Guessing, I installed it correctly, since scrolling is more reliable now

To be honest, I can't tell a difference after building and installing #3 !

**Comment by dev-soragoto** [comment]
> > Would really like to contribute but its not self explaining a lot and documentation is very bad regarding the parameters
> 
> Agreed, I find the documentation opaque and it doesn't help that the magic trackpad is a blackbox. It would be invaluable to see the process the previous devs used to build the original driver.
> 
> > A problem I encountered is this using libinput debug-events
> > event258 - Apple Inc. Magic Trackpad 2: kernel bug: Touch jump detected and discarded.
> > See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details
> 
> Interesting, I get this error too. I wonder if this occurs on older trackpads too. I went down a rabbit hole on a dmesg error `unknown main item tag 0x0` only to find logs online where this occurs with previous models.
> 
> > I am still confused because libinput measure touchpad-pressure prints out 253 as max value
> 
> I see the same thing 🙃. Even lowering it doesn't change what is reported by libinput, so I'm not sure whats going on there. Using evtest, libinput, etc, the trackpad appears extremely sensitive at detecting the lightest pressures of multiple fingers. If anything, I wonder if it could be when libinput decides to go from `POINTER_MOTION` to `POINTER_SCROLL_FINGER` and could be improved with libinput tuning and not at the driver level.
> 
> > Guessing, I installed it correctly, since scrolling is more reliable now
> 
> To be honest, I can't tell a difference after building and installing #3 !

After building and installing #3 , i still get this error. 

```bash
event21 - Apple Inc. Magic Trackpad USB-C: kernel bug: Touch jump detected and discarded.
See https://wayland.freedesktop.org/libinput/doc/1.27.0/touchpad-jumping-cursors.html for details
```

Interestingly, I have never encountered the issue of two-finger scrolling not working, no matter which version I use to build, I'm not sure why.

**Comment by dev-soragoto** [comment]
Anyway, I noticed this as well

```bash
    Event code 24 (ABS_PRESSURE)
      Value      0
      Min        0
      Max      253
    Event code 58 (ABS_MT_PRESSURE)
      Value      0
      Min        0
      Max      253
```

**Comment by dev-soragoto** [comment]
@SeDi343  I think this part of the code might need to be refactored

https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver/blob/1d798774dca415b7e2d81456e6bcb9290e58f836/linux/drivers/hid/hid-magicmouse.c#L643-L651

**Comment by SeDi343** [comment]
> > > Would really like to contribute but its not self explaining a lot and documentation is very bad regarding the parameters
> > 
> > 
> > Agreed, I find the documentation opaque and it doesn't help that the magic trackpad is a blackbox. It would be invaluable to see the process the previous devs used to build the original driver.
> > > A problem I encountered is this using libinput debug-events
> > > event258 - Apple Inc. Magic Trackpad 2: kernel bug: Touch jump detected and discarded.
> > > See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details
> > 
> > 
> > Interesting, I get this error too. I wonder if this occurs on older trackpads too. I went down a rabbit hole on a dmesg error `unknown main item tag 0x0` only to find logs online where this occurs with previous models.
> > > I am still confused because libinput measure touchpad-pressure prints out 253 as max value
> > 
> > 
> > I see the same thing 🙃. Even lowering it doesn't change what is reported by libinput, so I'm not sure whats going on there. Using evtest, libinput, etc, the trackpad appears extremely sensitive at detecting the lightest pressures of multiple fingers. If anything, I wonder if it could be when libinput decides to go from `POINTER_MOTION` to `POINTER_SCROLL_FINGER` and could be improved with libinput tuning and not at the driver level.
> > > Guessing, I installed it correctly, since scrolling is more reliable now
> > 
> > 
> > To be honest, I can't tell a difference after building and installing #3 !
> 
> After building and installing #3 , i still get this error.
> 
> ```shell
> event21 - Apple Inc. Magic Trackpad USB-C: kernel bug: Touch jump detected and discarded.
> See https://wayland.freedesktop.org/libinput/doc/1.27.0/touchpad-jumping-cursors.html for details
> ```
> 
> Interestingly, I have never encountered the issue of two-finger scrolling not working, no matter which version I use to build, I'm not sure why.

Yes I also get it with the Lightning Version. So it does not seem to be a problem specific with the USB-C Version.

```
event259 - Apple Inc. Magic Trackpad: kernel bug: Touch jump detected and discarded.
See https://wayland.freedesktop.org/libinput/doc/1.20.0/touchpad-jumping-cursors.html for details
```

-----

> To be honest, I can't tell a difference after building and installing #3 !

I am unsure if it helped for me using this code.
[https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver/blob/1d798774dca415b7e2d81456e6bcb9290e58f836/linux/drivers/hid/hid-magicmouse.c#L684-L690](https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver/blob/1d798774dca415b7e2d81456e6bcb9290e58f836/linux/drivers/hid/hid-magicmouse.c#L684-L690)

Or it was my sway config, including the parameter `left_handed true`:
```
input "76:804:Apple_Inc._Magic_Trackpad_USB-C" {
   dwt disabled
   tap disabled
   click_method clickfinger
   natural_scroll enabled
   #accel_profile "flat"
   accel_profile "adaptive"
   pointer_accel 0.5
   left_handed true
   middle_emulation disabled
}
```

-----

> @SeDi343 I think this part of the code might need to be refactored
> 
> https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver/blob/1d798774dca415b7e2d81456e6bcb9290e58f836/linux/drivers/hid/hid-magicmouse.c#L643-L651

True, or we find values to actually change the Pressure, since using
```
			input_set_abs_params(input, ABS_MT_PRESSURE, 0, 255, 0, 0);
			input_set_abs_params(input, ABS_PRESSURE, 0, 255, 0, 0);
```
did not change anything and I was still reading out max pressure of 253 in `libinput measure touchpad-pressure`.

-----

> I wonder if it could be when libinput decides to go from `POINTER_MOTION` to `POINTER_SCROLL_FINGER` and could be improved with libinput tuning and not at the driver level.

This might be the proper solution, if we can not really change the way the drivers translation of pressure on the trackpad.

**Comment by SeDi343** [comment]
I also noticed that the surface of the Trackpad is different to the Lightning version. If I have fresh washed hands, I stick more to the surface than on the Lightning Version. But there is nothing we can do about it :(

**Comment by Erma318** [comment]
> I think I am doing something wrong, since I see no difference in value `10-255` for `ABS_MT_PRESSURE` and `ABS_PRESSURE`. Whenever I make changes I run `./build-deb.sh` and then `sudo dpkg -I magicmouse-hid_2.1.0-0.deb`. And have Secure Boot active. Do I need to uninstall the driver first before installing it again?

So did you verified if we need to uninstall it before reinstall a dev version ?

**Comment by SeDi343** [comment]
Hey @Erma318, just building the deb using `./build-deb.sh` and installing it again with `sudo dpkg -I magicmouse-hid_2.1.0-0.deb` did not work. I used the uninstall `remove.sh` and install `install.sh` script present in the repository.

**Comment by Erma318** [comment]
> Hey @Erma318, just building the deb using `./build-deb.sh` and installing it again with `sudo dpkg -I magicmouse-hid_2.1.0-0.deb` did not work. I used the uninstall `remove.sh` and install `install.sh` script present in the repository.

Thanks! Will try!

**Comment by SeDi343** [comment]
> Any idea what `input_set_events_per_packet(input, 60);` stands for? Does it have something to do with the polling rate?

@mr-cal @dev-soragoto do you know what this function actually does?

**Comment by dev-soragoto** [comment]
> > Any idea what `input_set_events_per_packet(input, 60);` stands for? Does it have something to do with the polling rate?
> 
> @mr-cal @dev-soragoto do you know what this function actually does?

Maybe this function tell handlers about the driver event rate, I don't think it should be changed. changing it hasn't made any difference.

**Comment by SeDi343** [comment]
Does it have something to do with the polling rate? Since I think those Trackpads have 90-120Hz Polling Rate.

**Comment by SeDi343** [comment]
> @SeDi343 I think this part of the code might need to be refactored
> 
> https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver/blob/1d798774dca415b7e2d81456e6bcb9290e58f836/linux/drivers/hid/hid-magicmouse.c#L643-L651

Refactored this, to its original lines. Just kept `input_set_events_per_packet` update for now. Want to know what this really does.

**Comment by SeDi343** [comment]
@mr-cal and @dev-soragoto I think this rows could affect the multi touch recognition.
```
		pressure = tdata[7];
		state = tdata[3] & 0xC0;
		down = state == 0x80;
```
in
```
	} else if (input->id.product == USB_DEVICE_ID_APPLE_MAGICTRACKPAD2 ||
		   input->id.product ==
			   USB_DEVICE_ID_APPLE_MAGICTRACKPAD2_USBC) {
		id = tdata[8] & 0xf;
		x = (tdata[1] << 27 | tdata[0] << 19) >> 19;
		y = -((tdata[3] << 30 | tdata[2] << 22 | tdata[1] << 14) >> 19);
		size = tdata[6];
		orientation = (tdata[8] >> 5) - 4;
		touch_major = tdata[4];
		touch_minor = tdata[5];
		pressure = tdata[7];
		state = tdata[3] & 0xC0;
		down = state == 0x80;
```
https://github.com/SeDi343/Linux-Magic-Trackpad-2-USB-C-Driver/blob/1d798774dca415b7e2d81456e6bcb9290e58f836/linux/drivers/hid/hid-magicmouse.c#L242-L244

**Comment by SeDi343** [comment]
Sadly it seems like the info of Datalayout of Bluetooth packets seems to be outdated on https://github.com/RicardoEPRodrigues/magicmouse-hid

		/* The data layout for magic mouse 2 is:
		 * 14 bytes of prefix
		 * data[0] is the device report ID
		 * data[1] is the mouse click events. Value of 1 is left, 2 is right.
		 * data[2] (contains lsb) and data[3] (msb) are the x movement
		 *         of the mouse 16bit representation.
		 * data[4] (contains msb) and data[5] (msb) are the y movement
		 *         of the mouse 16bit representation.
		 * data[6] data[13] are unknown so far. Need to decode this still
		 *
		 * data[14] onwards represent touch data on top of the mouse surface
		 *          touchpad. There are 8 bytes per finger. e.g:
		 * data[14]-data[21] will be the first finger detected.
		 * data[22]-data[29] will be finger 2 etc.
		 * these sets of 8 bytes are passed in as tdata to
		 * magicmouse_emit_touch()
		 *
		 * npoints is the number of fingers detected.
		 * size is minimum 14 but could be any multpiple of 14+ii*8 based on
		 * how many fingers are detected. e.g for 1 finger, size=22 for
		 * 2 fingers, size=30 and so on.
		 */

        /* tdata is 8 bytes per finger detected.
		 * tdata[0] (lsb of x) and least sig 4bits of tdata[1] (msb of x)
		 *          are x position of touch on touch surface.
		 * tdata[1] most sig 4bits (lsb of y) and and tdata[2] (msb of y)
		 *          are y position of touch on touch surface.
		 * tdata[1] bits look like [y y y y x x x x]
		 * tdata[3] touch major axis of ellipse of finger detected
		 * tdata[4] touch minor axis of ellipse of finger detected
		 * tdata[5] contains 6bits of size info (lsb) and the two msb of tdata[5]
		 *          are the lsb of id: [id id size size size size size size]
		 * tdata[6] 2 lsb bits of tdata[6] are the msb of id and 6msb of tdata[6]
		 *          are the orientation of the touch. [o o o o o o id id]
		 * tdata[7] 4 msb are state. 4lsb are unknown.
		 *
		 * [ x x x x x x x x ]
		 * [ y y y y x x x x ]
		 * [ y y y y y y y y ]
		 * [touch major      ]
		 * [touch minor      ]
		 * [id id s s s s s s]
		 * [o o o o o o id id]
		 * [s s s s | unknown]
		 */

**Comment by SeDi343** [comment]
This is very strange, I tried to reverse engineer the x and y data to find out, that x and y are different size.

```
>>> tdata = (0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF, 0x5555555555555555, 0x6666666666666666, 0x7777777777777777, 0x8888888888888888, 0x9999999999999999)
>>> bin((tdata[3] << 30 | tdata[2] << 22 | tdata[1] << 14) >> 19)
'0b111111111111111111111111111111111111111111111111111111111111111111111111111'
>>> bin((tdata[1] << 27 | tdata[0] << 19) >> 19)
'0b111111111111111111111111111111111111111111111111111111111111111111111111'
>>> bin((-(tdata[3] << 30 | tdata[2] << 22 | tdata[1] << 14)) >> 19)
'-0b1000000000000000000000000000000000000000000000000000000000000000000000000000'
>>> bin((tdata[1] << 27 | tdata[0] << 19) >> 19)
'0b111111111111111111111111111111111111111111111111111111111111111111111111'
```
am I doing something wrong?

**Comment by SeDi343** [comment]
After digging more into the driver and debugging the events https://github.com/mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver/pull/3 I get more and more the feeling this might be more and more a problem with [libinput](https://gitlab.freedesktop.org/libinput/libinput).

Can you please investigate into the debug of `sudo journalctl --all --follow`  and `libinput debug-events` as well @mr-cal?

**Comment by SeDi343** [comment]
This is how the current Bluetooth Data looks like based on the input from the Driver. ? Is unknown data.

```
[ x x x x x x x x ]
[ y y y x x x x x ]
[ y y y y y y y y ]
[ s s ? ? ? ? y y ]
[ touch_major     ]
[ touch_minor     ]
[ size            ]
[ pressure        ]
[ o o o ? i i i i ]
```

**Comment by SeDi343** [comment]
> After digging more into the driver and debugging the events #3 I get more and more the feeling this might be more and more a problem with [libinput](https://gitlab.freedesktop.org/libinput/libinput).
> 
> Can you please investigate into the debug of `sudo journalctl --all --follow` and `libinput debug-events` as well @mr-cal?

Need to change `#define DEBUG 0` to `1` for debug messages on journalctl

**Comment by SeDi343** [comment]
@mr-cal can you submitting the new update of #3 as a patch to the kernel? I do not know, how to do this.

Greetings,
Sebastian

**Comment by mr-cal** [comment]
> @mr-cal can you submitting the new update of #3 as a patch to the kernel (Maybe when #6 is merged)? I do not know, how to do this.
> 
> Greetings, Sebastian

@SeDi343, I'm worried that getting rejected by the kernel maintainers because it's not entirely complete, mostly around the unknown contents of the data packet.

**Comment by SeDi343** [comment]
> > @mr-cal can you submitting the new update of #3 as a patch to the kernel (Maybe when #6 is merged)? I do not know, how to do this.
> > Greetings, Sebastian
> 
> @SeDi343, I'm worried that getting rejected by the kernel maintainers because it's not entirely complete, mostly around the unknown contents of the data packet.

But the unknown cotents of the data packet were already present before my changes. I just observed that those data values are not touched at all by the driver.

**Comment by SeDi343** [comment]
Hey @mr-cal wishing you a Happy new Year

Can you tell me how I could push patches to the kernel maintainers? Would like to try it.

**Comment by martinstark** [comment]
I've tested the latest source code in this repo and I'm still having the same issues with input dropping. It's immediately apparent when using Magic Trackpad 2 and the USB-C variant side-by-side.

The USB-C trackpad is super sensitive to touches (possibly more so than v2), but input still sometimes drops during multi gestures: two finder scrolling suddenly stops detecting one finger and instead moves the mouse, creating a choppy scrolling experience.

I don't necessarily think this is a force or sensitivity issue.

**Comment by mr-cal** [comment]
> Hey @mr-cal wishing you a Happy new Year
> 
> Can you tell me how I could push patches to the kernel maintainers? Would like to try it.

Absolutely, I think you should!

These are the guides I followed, I recommend using `git send-email`, it's the most foolproof method.
*  https://docs.kernel.org/process/submitting-patches.html
* https://kernelnewbies.org/FirstKernelPatch

**Comment by SeDi343** [comment]
> > Hey @mr-cal wishing you a Happy new Year
> > Can you tell me how I could push patches to the kernel maintainers? Would like to try it.
> 
> Absolutely, I think you should!
> 
> These are the guides I followed, I recommend using `git send-email`, it's the most foolproof method.
> 
>     * https://docs.kernel.org/process/submitting-patches.html
> 
>     * https://kernelnewbies.org/FirstKernelPatch

Thank you for the Information :)

**Comment by SeDi343** [comment]
> > I've tested the latest source code in this repo and I'm still having the same issues with input dropping. It's immediately apparent when using Magic Trackpad 2 and the USB-C variant side-by-side.
> > The USB-C trackpad is super sensitive to touches (possibly more so than v2), but input still sometimes drops during multi gestures: two finder scrolling suddenly stops detecting one finger and instead moves the mouse, creating a choppy scrolling experience.
> > I don't necessarily think this is a force or sensitivity issue.
> 
> Yeah, it seems like something is still slightly off but we haven't been able to figure it out yet. In #5, @SeDi343 has some logs with `libinput error: event259 - Apple Inc. Magic Trackpad: kernel bug: Touch jump detected and discarded.`. I've seen the same thing and libinput's doc's say it could be a driver issue but I can't figure it out.

Yes, I also get this with my Lightning Version. When tracking all Data with the debug option, all my touches are getting registered properly. So my guess would be nothing with the Driver more with the library above the driver.

**Comment by SeDi343** [comment]
> > > Hey @mr-cal wishing you a Happy new Year
> > > Can you tell me how I could push patches to the kernel maintainers? Would like to try it.
> > 
> > 
> > Absolutely, I think you should!
> > These are the guides I followed, I recommend using `git send-email`, it's the most foolproof method.
> > ```
> > * https://docs.kernel.org/process/submitting-patches.html
> > 
> > * https://kernelnewbies.org/FirstKernelPatch
> > ```
> 
> Thank you for the Information :)

Hope I did this correctly, added you in cc.

**Comment by SeDi343** [comment]
@martinstark  Try those steps as provided by @Erma318 in https://github.com/mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver/issues/5#issuecomment-2585464253.

This actually works for me.

**Comment by 123unix** [comment]
> I've tested the latest source code in this repo and I'm still having the same issues with input dropping. It's immediately apparent when using Magic Trackpad 2 and the USB-C variant side-by-side.

@martinstark do you mind mentioning the model of the v.2 trackpad please? (The **Axxxx** name along with the **Mxxxxxx/A** name)

Also, how would you compare the surfaces of the two? Do they feel the same or is one of them slightly less sticky?

Thanks!

**Comment by martinstark** [comment]
@123unix I can see a model number on the back of the trackpad: A3120, but I can't see anything corresponding to /A. Would it be at the back of the trackpad or do I need to check input in OS?

**Comment by SeDi343** [comment]
I think @123unix means means the EMC Number. For the USB C Version this should be 8538. Did you already install the quarks as mentioned in the [README.md - Configure Libinput](https://github.com/mr-cal/Linux-Magic-Trackpad-2-USB-C-Driver?tab=readme-ov-file#configure-libinput) Section.

I had the same problem as you. I was debugging the driver to its heart with the debug logs activated but did not find anything. Only after creating those quarks I was able to reliable work with the Trackpad.

**Comment by martinstark** [comment]
thanks for the tip, I had not noticed the README has been updated. Added the quirks and rebooted, seems good so far, but it's only been a minute 👍 

Should I close this issue if I find that this has fixed the problems? I'll use the usb c version for the rest of today to see if it's usable as a daily driver now

**Comment by mr-cal** [comment]
> Should I close this issue if I find that this has fixed the problems?

Yeah, I think if you're happy, we can close the issue. Anyone new who comes to this repo will see @SeDi343's updates to the README.

**Comment by martinstark** [comment]
With the quirks update the device has worked well for me for an hour of use, without it I would drop multi gesture inputs more or less immediately. I consider my reported problem fixed, thanks everyone 👍 

Is there an issue (or something) that I can follow in order to see if `libinput` is updated, in order to remove the quirks in the future?

**Comment by SeDi343** [comment]
@martinstark no there is none. I would welcome it, if you can open one if you want :)

**Comment by 123unix** [comment]
> [@123unix](https://github.com/123unix) I can see a model number on the back of the trackpad: A3120, but I can't see anything corresponding to /A. Would it be at the back of the trackpad or do I need to check input in OS?
> 
> I find the surface stickiness very similar

Yes, for USB-C version the model name is **A3120** and the other name or product ID is **MXK93AM/A** for white and **MXKA3AM/A** for black made in Vietnam (this can be seen in the URL of the apple.com shop page for this product and usually mentioned on other stores as well, e.g. on Amazon)

I was interested to know these names for the Lightning version you compared to this one.

**Comment by martinstark** [comment]
The lightning version is A1535 and EMC 2733, but I don't have the MX-number handy I'm afraid.
